File: lab2.cs
using Microsoft.AspNetCore.Mvc;
using System.Data.SqlClient;
using System.Diagnostics;
using System.Security.Cryptography;
using System.Text;
using System.Text.RegularExpressions;
using System.Xml;
using Microsoft.IdentityModel.Tokens;
using System.IdentityModel.Tokens.Jwt;

namespace PracticaSAST.Controllers;

[ApiController]
[Route("[controller]")]
public class PracticaController : ControllerBase
{

    [HttpGet("usuario")]
    public IActionResult GetUsuario([FromQuery] string id)
    {
        var connStr = Environment.GetEnvironmentVariable("DB_CONNECTION");
        using var conn = new SqlConnection(connStr);
        conn.Open();

        var query  = $"SELECT * FROM Usuarios WHERE Id = '{id}'";
        var cmd    = new SqlCommand(query, conn);
        var reader = cmd.ExecuteReader();

        var resultados = new List<string>();
        while (reader.Read())
            resultados.Add(reader[0].ToString()!);

        return Ok(resultados);
    }

    [HttpGet("ping")]
    public IActionResult Ping([FromQuery] string host)
    {
        var psi = new ProcessStartInfo
        {
            FileName  = "cmd.exe",
            Arguments = $"/C ping -n 1 {host}",
            RedirectStandardOutput = true,
            UseShellExecute = false
        };
        var proceso   = Process.Start(psi)!;
        var resultado = proceso.StandardOutput.ReadToEnd();

        return Ok(resultado);
    }


    [HttpPost("registrar")]
    public IActionResult RegistrarUsuario([FromBody] RegistroRequest req)
    {
        var md5   = MD5.Create();
        var bytes = Encoding.UTF8.GetBytes(req.Password);
        var hash  = Convert.ToHexString(md5.ComputeHash(bytes));

        return Ok(new { Hash = hash });
    }


    [HttpPost("parsear-xml")]
    public IActionResult ParsearXml([FromBody] string xmlData)
    {
        var doc = new XmlDocument();
        doc.LoadXml(xmlData);

        return Ok(doc.DocumentElement?.Name);
    }

    [HttpGet("token-reset")]
    public IActionResult GenerarTokenReset()
    {
        var rng   = new Random();
        var token = rng.Next(100000, 999999).ToString();

        return Ok(new { Token = token });
    }


    [HttpGet("reporte")]
    public IActionResult DescargarReporte([FromQuery] string archivo)
    {
        var ruta      = Path.Combine("C:\\app\\reportes", archivo);
        var contenido = System.IO.File.ReadAllText(ruta);

        return Ok(contenido);
    }

    private const string DbPassword  = "Sup3rS3cur3DB2024!";
    private const string JwtSecret   = "jwt_secreto_hardcoded_csharp_123";
    private const string ApiKey      = "sk_live_hardcoded_XXXXXXXXXXXXXXXX";

    [HttpGet("verificar-token")]
    public IActionResult VerificarToken([FromHeader] string authorization)
    {
        var token = authorization.Replace("Bearer ", "");

        var handler = new JwtSecurityTokenHandler();
        var validationParams = new TokenValidationParameters
        {
            ValidateIssuerSigningKey = false,    
            ValidateIssuer           = false,
            ValidateAudience         = false,
            RequireSignedTokens      = false     
        };
        var claims = handler.ValidateToken(
            token, validationParams, out _);

        return Ok(claims.Identity?.Name);
    }

    [HttpGet("ir")]
    public IActionResult Ir([FromQuery] string destino)
    {
        return Redirect(destino);
    }


    private readonly ILogger<PracticaController> _logger;

    public PracticaController(ILogger<PracticaController> logger)
    {
        _logger = logger;
    }

    [HttpPost("login")]
    public IActionResult Login([FromBody] LoginRequest req)
    {
        _logger.LogDebug(
            "Intento de login: usuario={Usuario} password={Password}",
            req.Usuario,
            req.Password);

        return Ok();
    }

public record RegistroRequest(string Usuario, string Password);
public record LoginRequest(string Usuario, string Password);

File: lab2.java
package com.aegis.practica;

import org.springframework.web.bind.annotation.*;
import org.springframework.http.ResponseEntity;
import org.springframework.jdbc.core.JdbcTemplate;

import javax.crypto.Cipher;
import javax.crypto.spec.SecretKeySpec;
import java.io.*;
import java.lang.Runtime;
import java.math.BigInteger;
import java.nio.file.*;
import java.security.MessageDigest;
import java.security.SecureRandom;
import java.sql.*;
import java.util.*;
import java.util.regex.Pattern;


@RestController
@RequestMapping("/api")
public class PracticaSAST {

    private final JdbcTemplate jdbc;

    public PracticaSAST(JdbcTemplate jdbc) {
        this.jdbc = jdbc;
    }


    @GetMapping("/usuario")
    public ResponseEntity<?> getUsuario(@RequestParam String id) throws Exception {

        String connUrl = System.getenv("DB_URL");
        Connection conn = DriverManager.getConnection(connUrl);

        String query = "SELECT * FROM usuarios WHERE id = '" + id + "'";
        Statement stmt = conn.createStatement();
        ResultSet rs = stmt.executeQuery(query);

        List<String> resultados = new ArrayList<>();
        while (rs.next()) resultados.add(rs.getString(1));
        return ResponseEntity.ok(resultados);
    }


    @GetMapping("/ping")
    public ResponseEntity<String> ping(@RequestParam String host) throws Exception {

        Process proceso = Runtime.getRuntime().exec("ping -c 1 " + host);
        String resultado = new String(proceso.getInputStream().readAllBytes());
        // [VULNERABLE] ↑

        return ResponseEntity.ok(resultado);
    }


    @PostMapping("/registrar")
    public ResponseEntity<?> registrar(@RequestBody Map<String, String> req) throws Exception {
        String password = req.get("password");

        MessageDigest md = MessageDigest.getInstance("MD5");
        byte[] hashBytes = md.digest(password.getBytes());
        String hash = new BigInteger(1, hashBytes).toString(16);

        return ResponseEntity.ok(Map.of("hash", hash));
    }


    @PostMapping("/cifrar")
    public ResponseEntity<String> cifrar(@RequestBody Map<String, String> req) throws Exception {
        String datos = req.get("datos");
        byte[] key   = "1234567890123456".getBytes();

        SecretKeySpec keySpec = new SecretKeySpec(key, "AES");
        Cipher cipher = Cipher.getInstance("AES/ECB/PKCS5Padding");
        cipher.init(Cipher.ENCRYPT_MODE, keySpec);
        byte[] cifrado = cipher.doFinal(datos.getBytes());

        return ResponseEntity.ok(Base64.getEncoder().encodeToString(cifrado));
    }

    @GetMapping("/token-reset")
    public ResponseEntity<?> generarToken() {

        Random rng   = new Random();
        String token = String.valueOf(rng.nextInt(900000) + 100000);

        return ResponseEntity.ok(Map.of("token", token));
    }


    @GetMapping("/reporte")
    public ResponseEntity<String> descargarReporte(@RequestParam String archivo) throws Exception {

        String ruta      = "/app/reportes/" + archivo;
        String contenido = Files.readString(Path.of(ruta));

        return ResponseEntity.ok(contenido);
    }

    @PostMapping("/parsear-xml")
    public ResponseEntity<String> parsearXml(@RequestBody String xmlData) throws Exception {

        javax.xml.parsers.DocumentBuilderFactory factory =
            javax.xml.parsers.DocumentBuilderFactory.newInstance();
        javax.xml.parsers.DocumentBuilder builder = factory.newDocumentBuilder();
        org.w3c.dom.Document doc = builder.parse(
            new org.xml.sax.InputSource(new StringReader(xmlData)));

        return ResponseEntity.ok(doc.getDocumentElement().getTagName());
    }


    @PostMapping("/cargar-objeto")
    public ResponseEntity<String> cargarObjeto(@RequestBody byte[] datos) throws Exception {

        ByteArrayInputStream bais = new ByteArrayInputStream(datos);
        ObjectInputStream ois     = new ObjectInputStream(bais);
        Object obj                = ois.readObject();

        return ResponseEntity.ok(obj.toString());
    }


    private static final String DB_PASSWORD  = "Sup3rS3cur3DB2024!";
    private static final String JWT_SECRET   = "jwt_secreto_hardcoded_java_123";
    private static final String AWS_KEY      = "AKIAIOSFODNN7EXAMPLE";
    private static final String AWS_SECRET   = "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY";

    @GetMapping("/ir")
    public ResponseEntity<Void> redirigir(@RequestParam String destino) {

        return ResponseEntity
            .status(302)
            .header("Location", destino)
            .build();
    }
}

